File: Beecrowd_Parte2/Infixa_para_Posfixa_1077.java
/*
O Professor solicitou que você escreva um programa que converta uma expressão na forma infixa (como usualmente conhecemos) para uma expressão na forma posfixa. Como você sabe, os termos in (no meio) e pos (depois) se referem à posição dos operadores. O programa terá que lidar somente com operadores binários +,-,*,/,^, parênteses, letras e números. Um exemplo seria uma expressão como:
(A*B+2*C^3)/2*A. O programa deve converter esta expressão (infixa) para a expressão posfixa: AB*2C3^*+2/A*
 

Entrada
A primeira linha da entrada contém um valor inteiro N (N < 1000), que indica o número de casos de teste. Cada caso de teste a seguir é uma expressão válida na forma infixa, com até 300 caracteres.

Saída
Para cada caso, apresente a expressão convertida para a forma posfixa.

Exemplo de Entrada	
3
A*2
(A*2+c-d)/2
(2*4/a^b)/(2*c)

Exemplo de Saída
A2*
A2*c+d-2/
24*ab^/2c*/

// OBS.:  O " */ "  de 24*ab^/2c*/ faz parte da saída

import java.util.Scanner;


public class Infixa_para_Posfixa_1077 {

    // Classe para representar o nó da pilha
    private static class PilhaNo {
        char valor;         // Valor armazenado no nó da pilha
        PilhaNo abaixo;     // Referência para o próximo nó abaixo na pilha

        PilhaNo(char valor) {
            this.valor = valor;
            this.abaixo = null;
        }
    }

    // Classe para representar a pilha
    private static class Pilha {
        int tamanho;        // Tamanho atual da pilha (quantidade de elementos)
        PilhaNo topo;       // Referência para o topo da pilha (último elemento inserido)

        Pilha() {
            this.tamanho = 0;
            this.topo = null;
        }

        void push(char valor) {
            PilhaNo novoTopo = new PilhaNo(valor);  // Cria um novo nó para o valor
            novoTopo.abaixo = topo;                 // Aponta o novo nó para o antigo topo
            topo = novoTopo;                        // Atualiza o topo da pilha
            tamanho++;                              // Aumenta o tamanho da pilha
        }

        void pop() {
            if (tamanho > 0) {
                topo = topo.abaixo;  // Atualiza o topo para o próximo nó abaixo
                tamanho--;           // Reduz o tamanho da pilha
            }
        }

        char top() {
            return topo.valor;  // Retorna o valor do topo da pilha
        }


        boolean isEmpty() {
            return tamanho == 0;  // Verifica se a pilha está vazia
        }
    }

    // Função para determinar a precedência de operadores
    private static int precedencia(char operador) {
        switch (operador) {
            case '+':
            case '-':
                return 1;   // Baixa precedência para soma e subtração
            case '*':
            case '/':
                return 2;   // Alta precedência para multiplicação e divisão
            default:
                return 0;   // Outros operadores têm precedência zero (ou não são operadores válidos)
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int N = scanner.nextInt();  // Lê o número de expressões
        scanner.nextLine();         // Consumir o newline pendente após nextInt

        for (int k = 0; k < N; ++k) {
            String expressao = scanner.nextLine();  // Lê a expressão matemática

            Pilha pilha = new Pilha();  // Cria uma nova pilha para cada expressão

            int tam = expressao.length();
            for (int i = 0; i < tam; ++i) {
                char caractere = expressao.charAt(i);

                if (Character.isLetterOrDigit(caractere)) {
                    System.out.print(caractere);  // Se for caractere alfabético ou dígito, imprime diretamente
                } else if (caractere == '(' || caractere == '^') {
                    pilha.push(caractere);  // Se for '(' ou '^', empilha na pilha
                } else if (caractere == ')') {
                    while (!pilha.isEmpty() && pilha.top() != '(') {
                        System.out.print(pilha.top());  // Desempilha e imprime até encontrar '('
                        pilha.pop();
                    }

                    if (!pilha.isEmpty()) pilha.pop();  // Remove '(' da pilha, se presente
                } else {
                    while (!pilha.isEmpty() && pilha.top() != '(' && precedencia(caractere) <= precedencia(pilha.top())) {
                        System.out.print(pilha.top());  // Desempilha e imprime enquanto a precedência for maior ou igual
                        pilha.pop();
                    }

                    pilha.push(caractere);  // Empilha o operador atual
                }
            }

            while (!pilha.isEmpty()) {
                System.out.print(pilha.top());  // Desempilha e imprime qualquer operador restante na pilha
                pilha.pop();
            }

            System.out.println();  // Nova linha para a próxima expressão
        }

        scanner.close();
    }
}

File: Beecrowd_Parte2/Infixa_para_Posfixa_1077.c
/*
O Professor solicitou que você escreva um programa que converta uma expressão na forma infixa (como usualmente conhecemos) para uma expressão na forma posfixa. Como você sabe, os termos in (no meio) e pos (depois) se referem à posição dos operadores. O programa terá que lidar somente com operadores binários +,-,*,/,^, parênteses, letras e números. Um exemplo seria uma expressão como:
(A*B+2*C^3)/2*A. O programa deve converter esta expressão (infixa) para a expressão posfixa: AB*2C3^*+2/A*
 

Entrada
A primeira linha da entrada contém um valor inteiro N (N < 1000), que indica o número de casos de teste. Cada caso de teste a seguir é uma expressão válida na forma infixa, com até 300 caracteres.

Saída
Para cada caso, apresente a expressão convertida para a forma posfixa.

Exemplo de Entrada	
3
A*2
(A*2+c-d)/2
(2*4/a^b)/(2*c)

Exemplo de Saída
A2*
A2*c+d-2/
24*ab^/2c*/

// OBS.:  O " */ "  de 24*ab^/2c*/ faz parte da saída


#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>

// Define a estrutura para o nó da pilha
typedef struct pilhaNo {
    char valor;             // Valor armazenado no nó da pilha
    struct pilhaNo* abaixo; // Ponteiro para o próximo nó abaixo na pilha
} PilhaNo;                  // typedef para simplificar referências futuras

// Define a estrutura para a pilha
typedef struct pilha {
    int tamanho;    // Tamanho atual da pilha (quantidade de elementos)
    PilhaNo* topo;  // Ponteiro para o topo da pilha (último elemento inserido)
} Pilha;            // typedef para simplificar referências futuras

// Função para inserir um novo elemento no topo da pilha
void push(Pilha* p, char valor) {
    p->tamanho += 1;                        // Aumenta o tamanho da pilha
    PilhaNo* novoTopo = (PilhaNo*) malloc(sizeof(PilhaNo));  // Aloca memória para um novo nó da pilha

    novoTopo->valor = valor;                // Define o valor do novo nó
    novoTopo->abaixo = p->topo;             // Aponta o novo nó para o antigo topo da pilha
    p->topo = novoTopo;                     // Atualiza o topo da pilha para o novo nó
}

// Função para remover o elemento do topo da pilha
void pop(Pilha* p) {
    if (p->tamanho > 0) {                   // Verifica se a pilha não está vazia
        p->tamanho -= 1;                    // Reduz o tamanho da pilha
        PilhaNo* velhoTopo = p->topo;       // Armazena o endereço do antigo topo da pilha
        p->topo = p->topo->abaixo;          // Atualiza o topo da pilha para o nó abaixo do antigo topo
        free(velhoTopo);                    // Libera a memória do antigo topo da pilha
    }
}

// Função para obter o valor do topo da pilha sem remover
char top(Pilha* p) {
    return p->topo->valor;  // Retorna o valor do topo da pilha
}

// Função para obter o tamanho atual da pilha
int size(Pilha* p) {
    return p->tamanho;  // Retorna o tamanho da pilha
}

// Função para verificar se a pilha está vazia
int empty(Pilha* p) {
    return p->tamanho == 0;  // Retorna verdadeiro se a pilha estiver vazia, falso caso contrário
}

// Função para inicializar a pilha
void inicializa(Pilha* p) {
    p->tamanho = 0;     // Inicializa o tamanho da pilha como zero
    p->topo = NULL;     // Define o topo da pilha como nulo
}

// Função para destruir a pilha liberando toda a memória alocada
void destroi(Pilha* p) {
    while (!empty(p)) {  // Enquanto a pilha não estiver vazia
        pop(p);          // Remove o topo da pilha
    }
}

// Função para determinar a precedência de operadores
int precedencia(char operador) {
    switch (operador) {
        case '+':
        case '-':   return 1;  // Baixa precedência para soma e subtração
        case '*':
        case '/':   return 2;  // Alta precedência para multiplicação e divisão
        default:    return 0;  // Outros operadores têm precedência zero (ou não são operadores válidos)
    }
}

int main() {
    int N, tam;
    Pilha p;
    char expressao[301];

    scanf("%d\n", &N);  // Lê o número de expressões

    for (int k = 0; k < N; ++k) {
        scanf("%s\n", expressao);  // Lê a expressão matemática

        inicializa(&p);  // Inicializa a pilha para cada nova expressão

        tam = strlen(expressao);  // Determina o tamanho da expressão
        for (int i = 0; i < tam; ++i) {
            if (isalpha(expressao[i]) || isdigit(expressao[i])) {
                printf("%c", expressao[i]);  // Se for caractere alfabético ou dígito, imprime diretamente
            } else if (expressao[i] == '(' || expressao[i] == '^') {
                push(&p, expressao[i]);  // Se for '(' ou '^', empilha na pilha
            } else if (expressao[i] == ')') {
                while (!empty(&p) && top(&p) != '(') {
                    printf("%c", top(&p));  // Desempilha e imprime até encontrar '('
                    pop(&p);
                }

                if (!empty(&p))  pop(&p);  // Remove '(' da pilha, se presente
            } else {
                while (!empty(&p) && top(&p) != '(' && precedencia(expressao[i]) <= precedencia(top(&p))) {
                    printf("%c", top(&p));  // Desempilha e imprime enquanto a precedência for maior ou igual
                    pop(&p);
                }

                push(&p, expressao[i]);  // Empilha o operador atual
            }
        }

        while (!empty(&p)) {
            printf("%c", top(&p));  // Desempilha e imprime qualquer operador restante na pilha
            pop(&p);
        }
        printf("\n");  // Nova linha para a próxima expressão
    }

    return 0;
}
